### DISCUSSION TITLE
Multiple screens using the same but different providers

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#1362] [Q&A] [answered]
EDIT: “Optimal solution” can be found in this comment: https://github.com/rrousselGit/riverpod/discussions/1362#discussioncomment-2578197

Let’s say, I have an app that has a single search screen with a TextField search bar and a ListView with search results.

I am using a StateProvider to manage the current search query and a depending FutureProvider to manage the async search results. The screen is implemented using ConsumerWidget to access the providers.

However, now I want to stack the screens. Whenever the user taps a word in any of the ListView entries, I want to push another search screen that has the tapped word as search phrase. And that screen has a back button to dismiss it. But if the user again taps a word, yet another screen is pushed. (Recursive)

So, the problem now is that with my single provider, I can’t manage independent state across screens of the same kind.

What would be the preferred way to solve that problem?

**Comment by rrousselGit** [comment, marked as answer]
You could use family or scoping

Generally I'd recommend avoiding relying too much on scoping. But for passing a simple variable around, that's fine

Family though isnt a problem. It'll nicely support having multiple search at the same time

**Reply by derolf** [reply]
Of course, my example was oversimplified, but it basically expands to any app that has a kind of stack of similar pages with their own local state.

My current solution is to have a class with a bunch of providers inside. I am instantiating that class and pass it around. Basically a kind of scoping.

**Reply by rrousselGit** [reply]
Don't do that.

The solutions I proposed are the official ones. Definitely try to use family

**Reply by derolf** [reply]
If I would use family, I would need a key to locate the scoped state in the families. In that case I would need to pass the key around. Kinda same same? 

In theory, the key could be the ModalRoute enclosing the screen…?

**Reply by rrousselGit** [reply]
The idea is to use scoped providers to pass the family key.

So you end-up with:

```dart
Widget build(context, ref) {
  final key = ref.watch(scopedFamilyKey);
  final value = ref.watch(family(key));
}
```

You can simplify that by making a provider for it:

```dart
final defaultFamily = Provider((ref) {
  final key = ref.watch(scopedFamilyKey);
  final value = ref.watch(family(key));
}, dependencies: [scopedFamilyKey, family]);

ref.watch(defaultFamily)
```

**Reply by derolf** [reply]
Interesting concept, so a scoped provider for the key and a family to maintain individual states.

Great idea!!!!!

**Reply by rrousselGit** [reply]
Indeed! This way the family isn't scoped, so if you somehow still need to access the family outside of the scope, you can
And you still support multiple pages with different values.

**Reply by derolf** [reply]
Still a question: is it necessary to pass dependencies to the Provider? I thought dependencies are managed automatically when using ref.watch?

**Reply by rrousselGit** [reply]
`dependencies` is for automatically scoping that provider when one of its dependencies is scoped.

ref.watch does not support this

**Reply by droidpl** [reply]
@rrousselGit I still don't fully grasp how this solution works. For this use case I was using family and relying on passing an "id" with `Equatable` on each of the screens to make it work, but that solution seemed hacky to me.

Can you elaborate a bit more on the example above? 🙄 How that would work for example with a search screen? Thanks a lot for your work on this library

**Reply by derolf** [reply]
@droidpl I am doing it exactly that way. Each of my screens that should have independent state gets something that identifies it uniquely and then I use families all over the place.

**Reply by droidpl** [reply]
Yeah but for me this is forcing to generate the id in each screen, making the widget stateful and forcing me have extra unneeded properties for each of the dependent providers for those screens.

I was wondering if I was missing something with the dependencies property that I am not using and this family provider and scoped key proposed.

I think having a quick minimal example would help to see how would ve the cleanest possible way to do it 😊 maybe something we could have also in the docs?

**Reply by derolf** [reply]
I am not generating the ID IN the screen, but pass the ID from the outside TO the screen as a parameter to the constructor.

As a proxy you could use ModalRoute.of(context) as your key.
